data-generation/src/main/scala/thesis/SnapshotDelta.scala

```diff
@@ -68,8 +68,29 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
       forwardApplyLogs(closestGraph.graph, logsToApply)
     }
   }
-  
-  override def directNeighbours(vertexId: VertexId, interval: Interval): RDD[VertexId] = throw new NotImplementedError()
+
+  override def directNeighbours(vertexId: VertexId, interval: Interval): RDD[VertexId] = {
+    val filteredLogs = logs.filter(log => log.action == CREATE || log.action == DELETE)
+      .filter(log => log.entity match {
+        case VERTEX(_) => false
+        case EDGE(_, srcId, dstId) => srcId == vertexId || dstId == vertexId
+      })
+
+    LogUtils.getEdgeLogsById(filteredLogs)
+      .map(log => log._2.size match { // The corresponding actions should either be (CREATE) or (CREATE+DELETE)
+        case 1 => (log._2.toSeq(0).entity match { // The action is (CREATE) - the edge is never deleted
+          case VERTEX(_) => throw new EntityFilterException
+          case EDGE(_, srcId, dstId) => if (vertexId == srcId) dstId else srcId
+        }, Interval(log._2.toSeq(0).timestamp, Instant.MAX))
+        case 2 => (log._2.toSeq(0).entity match { // The actions are (CREATE+DELETE) - the edge is deleted
+          case VERTEX(_) => throw new EntityFilterException
+          case EDGE(_, srcId, dstId) => if (vertexId == srcId) dstId else srcId
+        }, Interval(log._2.toSeq(0).timestamp, log._2.toSeq(1).timestamp))
+        case _ => throw new IllegalStateException(s"Only CREATE or CREATE+DELETE allowed here.")
+      })
```

Comment:
Hvis du vil være helt vill, og litt mer deklarativ kan du gjøre dette:
```suggestion
      .map(log => log._2.map(_.action) match {
        case CREATE :: Nil => (log._2.toSeq(0).entity match { 
          case VERTEX(_) => throw new EntityFilterException
          case EDGE(_, srcId, dstId) => if (vertexId == srcId) dstId else srcId
        }, Interval(log._2.toSeq(0).timestamp, Instant.MAX))
        case CREATE::DELETE::Nil => (log._2.toSeq(0).entity match { //We assume the the logs are sorted and that DELETE::CREATE can not happen
          case VERTEX(_) => throw new EntityFilterException
          case EDGE(_, srcId, dstId) => if (vertexId == srcId) dstId else srcId
        }, Interval(log._2.toSeq(0).timestamp, log._2.toSeq(1).timestamp))
        case _ => throw new IllegalStateException(s"Only CREATE or CREATE+DELETE allowed here.")
      })
```
Da sier du også eksplisitt hvilke actions du forventer, hvor det tydeligere kun var implisitt.

Reply:
Ah det er myyyye nicere, takk!